Lib/csv.py

```diff
@@ -364,31 +364,35 @@ def _guess_delimiter(self, data, delimiters):
         try and evaluate the smallest portion of the data possible, evaluating
         additional chunks as necessary.
         """
+        from collections import Counter, defaultdict
 
         data = list(filter(None, data.split('\n')))
 
-        ascii = [chr(c) for c in range(127)] # 7-bit ASCII
-
         # build frequency tables
         chunkLength = min(10, len(data))
         iteration = 0
-        charFrequency = {}
+        seen = 0
+        # {char -> {count_per_line -> num_lines_with_that_count}}
+        charFrequency = defaultdict(Counter)
         modes = {}
         delims = {}
         start, end = 0, chunkLength
         while start < len(data):
             iteration += 1
             for line in data[start:end]:
-                for char in ascii:
-                    metaFrequency = charFrequency.get(char, {})
-                    # must count even if frequency is 0
-                    freq = line.count(char)
-                    # value is the mode
-                    metaFrequency[freq] = metaFrequency.get(freq, 0) + 1
-                    charFrequency[char] = metaFrequency
-
-            for char in charFrequency.keys():
-                items = list(charFrequency[char].items())
+                seen += 1
+                charCounts = Counter(line)
+                for char, count in charCounts.items():
+                    if ord(char) < 127:
+                        charFrequency[char][count] += 1
+
+            for char, counts in charFrequency.items():
+                presentCount = sum(counts.values())
+                zeroCount = seen - presentCount
+                if zeroCount > 0:
+                    items = list(counts.items()) + [(0, zeroCount)]
+                else:
+                    items = list(counts.items())
```

Comment:
```suggestion
                items = list(counts.items())
                missed_lines = seen - sum(counts.values())
                if missed_lines:
                    # charFrequency[char][0] can only be deduced now
                    # as it cannot be obtained when parsing the lines.
                    assert 0 not in counts.keys()
                    # Store the number of lines 'char' was missing from.
                    items.append((0, missed_lines))
```

I think this is what you want right?

Reply:
I'm not entirely sure of my suggestion so someone else needs to check this (I'm sick hence tired..)

Reply:
@picnixz It’s OK. I’ve spent way too much thinking about this code and I believe they’re the same, yours a bit more readable. :-)

I’m running the benchmark as we speak. 

There are only three gotchas:

- Not sure if I agree with the comment (the zeros can be obtained but it’s inefficient),
- Are we OK with leaving the `assert`?
- There is a very subtle difference between the original code and my code, that is the tie break. I don’t think it matters, though. 

Get better soon!

Reply:
> Not sure if I agree with the comment (the zeros can be obtained but it’s inefficient),

Yes if we change the construction. But with the usage of `Counter(line)` it's not possible to obtain them as we ... don't count chars that are missing.

> There is a very subtle difference between the original code and my code, that is the tie break. It should not matter, though.

Oh. Can you give us an example if possible? and then we can add a test as well.

> Are we OK with leaving the assert?

This specific assert should be cheap but we can remove it (along with the comment that can be misread). It's good that benchmarks are performed with that assert though.

Reply:
> > There is a very subtle difference between the original code and my code, that is the tie break. It should not matter, though.
> 
> Oh. Can you give us an example if possible? and then we can add a test as well.

Definitely!

- https://github.com/python/cpython/pull/137628/files#diff-dd5e5efc1566b1c1dc988769a04cc403d2416c8af5bf4785bec235d02a780813R1447

> > Are we OK with leaving the assert?
> 
> This specific assert should be cheap but we can remove it (along with the comment that can be misread). It's good that benchmarks are performed with that assert though.

I've removed it:

- https://github.com/python/cpython/pull/137628/files#diff-e4e9ef49f9188f63d0c22e6e24cf6a4880fca56c11077df84f64937c9081e2eaR392-R395

The new benchmark:

- https://github.com/maurycy/cpython-test/tree/main/csv-sniffer-counter-set

ran over 4b62c848bae2e2a28e8d70a8aa00ea35b5f5b0ca:

- https://github.com/python/cpython/compare/main...4b62c848bae2e2a28e8d70a8aa00ea35b5f5b0ca

and included the `assert`.